### ISSUE TITLE
Is it possble to add a registerable api at the beginning of torch.save

### ISSUE THREAD

**Issue by CLiqing** [pytorch/pytorch#117840]
### 🚀 The feature, motivation and pitch

Similar to https://github.com/pytorch/pytorch/pull/99808, we may have some special data in our tensor. We hope our tensor can be loaded by other devices like ``cpu`` so we need to change our tensors before torch.save. 

<img src="https://github.com/pytorch/pytorch/assets/23721160/24312ba8-5689-44f6-bae2-87b5c0f8bd48.png" width="50%">

Now we can only use monkey patches to do that for the only registerable step in torch.save is after data's(storage's) save. 

![image](https://github.com/pytorch/pytorch/assets/23721160/0796b34f-9f46-4609-a506-850e3339581d)

I wonder if there can be a registerable api for ``privateuse1`` backend at the beginning of torch.save so that we can change our data from special struct to general form before saving the actual data.

### Alternatives

Add a registerable api at the beginning of torch.save, for examale ``prepare_for_save``

cc @mruberry @mikaylagawarecki

**Comment by malfet** [comment]
@jbschlosser is PrivateUse1 the right label for such issues?

**Comment by CLiqing** [comment]
@ezyang could u please take a look at this? Thanks!

**Comment by CLiqing** [comment]
During ``torch.save``, ``tensor`` and ``storage`` are saved separately.
While saving ``tensor``, ``storage_numel`` is saved as well as ``storage_key`` and so on.
https://github.com/pytorch/pytorch/blob/ee1dbb2acf3f9b304c5f47b919e8de0fe8fee45c/torch/serialization.py#L828-L832

After then, ``storage`` are converted to ``cpu`` and saved.
https://github.com/pytorch/pytorch/blob/ee1dbb2acf3f9b304c5f47b919e8de0fe8fee45c/torch/serialization.py#L858-L859

However, it means ``torch`` will save ``storage_numel`` which has some special data for our ``privateuse1 `` backend and larger and then, save ``storage``'s data which has been converted to ``cpu`` and its size is smaller. ``storage_numel`` and ``storage``'s data are contradictory.
So we want to add registerable apis and allow ``privateuse1`` backend to preprocessed the data before ``torch.save``.

**Comment by CLiqing** [comment]
> Can you explain to me how torch.save works today? IIUC, wouldn't we convert the tensor to cpu first before writing it out?

During ``torch.save``, ``tensor`` and ``storage`` are saved separately.
While saving ``tensor``, ``storage_numel`` is saved as well as ``storage_key`` and so on.
https://github.com/pytorch/pytorch/blob/ee1dbb2acf3f9b304c5f47b919e8de0fe8fee45c/torch/serialization.py#L828-L832

After then, ``storage`` are converted to ``cpu`` and saved.
https://github.com/pytorch/pytorch/blob/ee1dbb2acf3f9b304c5f47b919e8de0fe8fee45c/torch/serialization.py#L858-L862

However, it means ``torch`` will save ``storage_numel`` which has some special data for our ``privateuse1`` backend and larger than real data. Real data means the size of ``storage's data`` converted from ``privateuse1`` backend in above code.
 ``storage_numel`` and ``storage's data`` are not match and it will cause a bug while loading.
<img src="https://github.com/pytorch/pytorch/assets/23721160/24312ba8-5689-44f6-bae2-87b5c0f8bd48.png" width="50%">

So we want to add registerable apis and allow ``privateuse1`` backend to preprocessed the data before ``torch.save``.

**Comment by ezyang** [comment]
Naively, why doesn't the storage conversion to cpu preserve the metadata?

**Comment by CLiqing** [comment]
> Naively, why doesn't the storage conversion to cpu preserve the metadata?

Sorry, I may not have been clear at first. Our storage data may be larger than in the CPU. Here ``storage data`` refers to the part pointed to by ``data_ptr``. For performance, we may add some padding, for example, aligning the data to 16. Although we will restore it when converting back to ``CPU``, ``storage_numel`` has been saved before. They're not matched.

So we want to add registerable apis and allow our ``privateuse1`` backend to preprocessed the data (convert to cpu) at the beginning of torch.save, @ezyang

**Comment by ezyang** [comment]
Why not preserve the padding when you convert to cpu?

**Comment by CLiqing** [comment]
Because we want ``storage data`` we save to be read and processed correctly by other devices.
For example, if the size of ``storage data`` is 10 (on CPU), some padding may be added by our ``PrivateUse1`` device to change the size to 15. This padding is unique to our device. and  we hope we can still save the data of size 10. We will add paddings again when our device loads data. Other devices, like CPU or CUDA, just load the data of size 10 and don't need to consider about the special data struct.

**Comment by ezyang** [comment]
This doesn't seem like a big deal to me? If you have padding, but the tensor metadata (e.g., storage offset, size) is set appropriately, CPU tensor would handle it correctly. And if you actually convert to cpu e.g., with .cpu() we wouldn't preserve the padding in this case.

**Comment by CLiqing** [comment]
> This doesn't seem like a big deal to me? If you have padding, but the tensor metadata (e.g., storage offset, size) is set appropriately, CPU tensor would handle it correctly. And if you actually convert to cpu e.g., with .cpu() we wouldn't preserve the padding in this case.

1. However, we can't set metadata ``storage.nbytes()`` "appropriately". We should keep the actual value which includes the part of padding for it is used in many apis. We will get more errors if we keep it without the size of padding. 
https://github.com/pytorch/pytorch/blob/adff3350959dc247c161e358faa43432c883a7a4/torch/csrc/Storage.cpp#L484-L488
https://github.com/pytorch/pytorch/blob/adff3350959dc247c161e358faa43432c883a7a4/aten/src/ATen/native/TensorShape.cpp#L370-L374
2.When we convert to cpu, we wouldn't preserve the padding and want to use the value without padding.

At the present, we can only use patches to solve this mismatch problem like,
```
torch.storage.UntypedStorage.nbytes = get_nbytes_without_paddings
result = torch.serialization.save(obj, f, pickle_module, pickle_protocol, True, _disable_byteorder_record)
torch.storage.UntypedStorage.nbytes = get_nbytes
return result
```
We hope there can be one such api so that we can do some preprocessing to metadata.

**Comment by CLiqing** [comment]
https://github.com/pytorch/pytorch/issues/117840



> This doesn't seem like a big deal to me? If you have padding, but the tensor metadata (e.g., storage offset, size) is set appropriately, CPU tensor would handle it correctly. And if you actually convert to cpu e.g., with .cpu() we wouldn't preserve the padding in this case.

1. However, we can't set metadata ``storage.nbytes()`` "appropriately". We should keep the actual value which includes the part of padding for it is used in many apis. We will get more errors if we keep it without the size of padding. 
https://github.com/pytorch/pytorch/blob/adff3350959dc247c161e358faa43432c883a7a4/torch/csrc/Storage.cpp#L484-L488
2.When we convert to cpu, we wouldn't preserve the padding and want to use the value without padding.

At the present, we can only use patches to solve this mismatch problem like,
```
torch.storage.UntypedStorage.nbytes = get_nbytes_without_paddings
result = torch.serialization.save(obj, f, pickle_module, pickle_protocol, True, _disable_byteorder_record)
torch.storage.UntypedStorage.nbytes = get_nbytes
return result
```
We hope there can be one such api so that we can do some preprocessing to metadata. @ezyang

**Comment by CLiqing** [comment]
@ezyang , what do you think? Can you give me some suggestions?

**Comment by ezyang** [comment]
In CPU land, I can have a CPU tensor that points to a storage buffer with padding before/after the tensor data. In this case, I have a non-zero storage offset. When I save this tensor, by default I *include* the padding. You can observe this in the following example:

```
import torch

x = torch.randn(1024)
torch.save(x[512:], 'foo.pt')
print(torch.load('foo.pt').storage_offset())
```

Nothing you have told me thus far has therefore convinced me that what you want to do isn't expressible without a hook.

**Comment by CLiqing** [comment]
> In CPU land, I can have a CPU tensor that points to a storage buffer with padding before/after the tensor data. In this case, I have a non-zero storage offset. When I save this tensor, by default I _include_ the padding. You can observe this in the following example:
> 
> ```
> import torch
> 
> x = torch.randn(1024)
> torch.save(x[512:], 'foo.pt')
> print(torch.load('foo.pt').storage_offset())
> ```
> 
> Nothing you have told me thus far has therefore convinced me that what you want to do isn't expressible without a hook.

In fact, our padding is more complex. We may modify the structure of data in a variety of ways depending on the scenario, not just adding padding at the end.
![image](https://github.com/pytorch/pytorch/assets/23721160/2fd3f683-21dc-41cc-83b4-6ed1f6481e5b)

These are so complex that we don't want others using our data need to learn and understand. We hope we can restore data back to its original state when we use ``torch.save`` to save it. But there is no such an entrance now.

**Comment by ezyang** [comment]
The first padding scenario can be accurately expressed using strides which make you skip two blue widths.

The second padding scenario can also be expressed by using strides to express transposition, and also expanding the stride to skip past the blue padding.

I can give concrete code examples if you still don't get it.

**Comment by CLiqing** [comment]
> The first padding scenario can be accurately expressed using strides which make you skip two blue widths.
> 
> The second padding scenario can also be expressed by using strides to express transposition, and also expanding the stride to skip past the blue padding.
> 
> I can give concrete code examples if you still don't get it.

I know these can be expressed by some ways, but we do not intend to increase learning costs for our clients. We hope to maintain these specific operations implicitly. Regardless of which type of paddings we use, others don't need to add some skippings or strides explicitly.
In ``torch.save``, we convert data to CPU, remove paddings and save the special format in metadata. When it is loaded into our ``PrivateUse1`` device, we will convert it back with special paddings and when loading it into ``CPU``, there is no need to do anything , @ezyang

**Comment by CLiqing** [comment]
@ezyang, so what's ur opinions now?

**Comment by ezyang** [comment]
I don't think having accurate strides makes client code more complicated, IMO

**Comment by CLiqing** [comment]
> I don't think having accurate strides makes client code more complicated, IMO

@ezyang, you're familiar with torch so it's not complicated to you, but maybe it's not for our clients. We just want to hand over code adaptation work by our ourselves rather than clients. They only need to specify device as ``PrivateUse1`` to run the code which is available in ``CPU`` or ``CUDA``, instead of adding strides everywhere in the code.
```
# Client code in ``CUDA``
torch.cuda.set_device(i)
...
b = a.clone()
...


# Client code in ``PrivateUse1``
# Only a few times
torch.privateuse1.set_device(i)
...
# Many times
b = a[xxx:yyy].clone()
...
```
It will not affect existing torch logic, but merely provide an interface to ``PrivateUse1`` device.
```
def save(
    obj: object,
    f: FILE_LIKE,
    pickle_module: Any = pickle,
    pickle_protocol: int = DEFAULT_PROTOCOL,
    _use_new_zipfile_serialization: bool = True,
    _disable_byteorder_record: bool = False
) -> None:
    if prepare_for_save is not None:
        prepare_for_save(obj)
    ...
```

**Comment by ezyang** [comment]
In general, users are expected not to deal explicitly with strides. If I add two values with unusual striding, I will typically preserve this striding (but make the result contiguous). If you are doing work to automatically insert padding when you predict it is necessary, this would be the same as just making functions output non-standard strides differently than pytorch on cpu/cuda.

Look, if you tell me, "Edward, I understand this is hypothetically better, but we already wrote lots of code managing different storage padding at the storage level, and we'd have to refactor all of it which we don't have time to do" I'd be like, OK, fine, this seems like a decent tradeoff to make if you've painted yourself into this situation, but then, what's so bad about an extra monkey patch on top. If we're putting something into PyTorch core, we're going to do it right, because we're going to have to maintain it forever afterwards.
